esmtools/composite.py

```diff
@@ -0,0 +1,105 @@
+from .utils import check_xarray
+from .stats import ttest_ind_from_stats
+import warnings
+import xarray as xr
+
+
+def standardize(ds, dim='time'):
+    return (ds - ds.mean(dim)) / ds.std(dim)
+
+
+def _create_composites(anomaly_field, index, threshold=1, dim='time'):
+    """Creates composite from some variable's anomaly field and a climate
+    index"""
+    index_comp = xr.full_like(index, 'none', dtype='U4')
+    index_comp[index >= threshold] = 'pos'
+    index_comp[index <= -threshold] = 'neg'
+    composite = anomaly_field.groupby(
+        index_comp.rename('index'))
+    return composite
+
+
+@check_xarray([0, 1])
+def composite_analysis(field,
+                       index,
+                       threshold=1,
+                       plot=False,
+                       ttest=False,
+                       psig=0.05,
+                       **plot_kwargs):
+    """Create composite maps based on some variable's response to a climate
+    index.
+
+    Make sure that the field and index are detrended prior to using this
+    function if needed.
+
+    Args:
+        field (xr.object): Variable to create composites for. Contains dims
+                           `time` and 2 spatial.
+        index (xr.object): Climate index time series.
+        threshold (float): Threshold value for standardized composite.
+                           Defaults to 1.
+        plot (bool): Quick plot and no returns. Defaults to False.
+        ttest (bool): Apply `ttest` whether pos/neg different from mean.
+                      Defaults to False.
+        psig (float): Significance level for ttest. Defaults to 0.05.
+        **plot_kwargs (type): kwargs to pass to xarray's plot function.
+
+    Returns:
+        composite (xr.object): Positive and negative composite if `not plot`.
+
+    References:
+        * Motivated from Ryan Abernathy's notebook here:
+        https://rabernat.github.io/research_computing/xarray.html
+    """
+    def compute_ttest_for_composite(composite, index, psig):
+        """
+        Computes the ttest for the composite relative to neutral years and
+        returns a masked map based on some alpha level.
+
+        Args:
+            composite: The grouped composite object
+            index: 'pos' or 'neg'
+            psig: Significance level for ttest
+        """
+        # Suppress NaN of empty slice warning.
+        with warnings.catch_warnings():
```

Comment:
Would like to make this a decorator. I think decorators can be passed within functions? Essentially just want to be able to wrap a block of code with this so we don't have to put it in three spots here.

Reply:
dont know. I didnt really understand the decorators yet to the fullest extent

Reply:
Since I don't plan on `esmtools` to ever be truly "released" we are fine with some sloppy code in some places. I.e., a few repetitive warnings. I might fix this at one point just to learn.